### ISSUE TITLE
Multi-threading problem

### ISSUE THREAD

**Issue by noembryo** [VeNoMouS/cloudflare-scrape-js2py#26]
OK I know I'm asking for too much, but I had to :o)

I'm using your great script in an [app of mine (Subber)](http://www.noembryo.com/apps.php?subber), because recently Subscene.com started to use CloudFlare.
And although it works most of the time while checking for a video, trying for more than one using different threads returns a ton of errors and of course no subtitles.

Is this a problem with js2py not being thread-safe, or something that I can fix within my code?

**Comment by pannal** [comment]
Just a guess: CF might not allow multiple solves at a time for the same page? (You wouldn't do that to a subtitle provider anyways, would you?)

**Comment by noembryo** [comment]
I create a scraper for every video (not every subtitle) but now that you mention it, I could use one scraper for every session!!
I'll try that and report back..

**Comment by noembryo** [comment]
No luck :o( 
Its the same result.

When the search starts,
I create a scraper (with `cfscrape.create_scraper()`)
and use it to first get the results page for every video `result = self.scraper.get(url).content` ,
and then use it to get extra info for every subtitle of the resulting pages.
All these are in different threads.

If I check for one video, it works OK.
If more than one, most of the time does not return anything, or sometimes returns results for one of the videos, or maybe two, but never more..

**Comment by pannal** [comment]
Yeah, don't create multiple sessions from one IP to the same site I guess.
If you're managing sessions, you might want to pass your original session object as `sess` to `create_scraper` to avoid creating new sessions.

**Comment by noembryo** [comment]
But I use the same scraper object for all the calls..

**Comment by pannal** [comment]
Use the same session for all the calls.

**Comment by noembryo** [comment]
How can I do that?

**Comment by pannal** [comment]
By creating a session beforehands, then passing it to the create_scraper calls. Or subclass CloudflareScraper in the first place.

**Comment by noembryo** [comment]
Just tried sess (`sess = cfscrape.create_scraper()`) with the same results..

**Comment by pannal** [comment]
Are you doing subsequent calls without using self.scraper perhaps? Can I see your code?

**Comment by noembryo** [comment]
No other calls are made.
The code spans some scripts.
I'll try to post some parts here..

**Comment by noembryo** [comment]
Hmm.. Posting only the relevant parts is not going to make any sense.
And I can't post it all, because is not open source.. :o(

But I'm almost certain that all the calls are made with the same object.
When I search for one video it works.
It also uses many threads (for the subtitles)

But if I search for more videos (for the initial results page), nothing returns (usually, not always)..

**Comment by pannal** [comment]
Nothing returns? Do you pause for a couple of seconds between requests?

**Comment by noembryo** [comment]
The calls are in different threads.
I can't pause between them.
They ask in parallel.

I was using the `request` lib before, and I now replace it with the scraper.

I'll try to block them until the first one get something...

**Comment by pannal** [comment]
> Just a guess: CF might not allow multiple solves at a time for the same page?

**Comment by noembryo** [comment]
Success..
After the first call, I block all the other threads until I have the first response.
After that, I let them call as they like and the results are the expected (most of the time)

The only problem is that even the js is solved, it takes a lot longer than the requests calls I used before the CloudFlare problem.

Anyway, beggars can't be choosers ;o)
Thank you @pannal for your help..

**Comment by VeNoMouS** [comment]
Glad you got it working guys... personally its 

```
import requests
from collection import OrderedDict
session = requests.session()
session.headers = OrderedDict()
cf = cfscrape.create_scraper(sess=session)
cf.get('http://somepagewithIUAM')
session.get('whatever')
```

if you were using sessions..., what happens is, __cfscrape__ takes the session pointer, does the __Challenge Response__, gets the __cfuid__ cookie and populates the __session cookie__ and returns ... then you can use normal requests ... or you can keep using __cf.get()__ which is just a __super()__ to requests but some actionis in the middle to detect the __Challenge Request__...

Just so your aware @pro-src and myself have been working on a new project that will supersede cfscrape... and personally is way better than this existing project.

**Comment by noembryo** [comment]
Is the `cf.get('http://TheWebSite')` and `session.get('http://TheWebSite')` return the same thing?

Is my cration of `self.scraper = cfscrape.create_scraper()` and using this object to get `result = self.scraper.get(url).content` for all the calls, any different?

Also, do I have to reset something somewhere between sessions, because some times I don't get a result at all?

Its great that you work on creating an even better solution!!

**Comment by VeNoMouS** [comment]
It will return the same thing the `cf.get()` will solve the challenge if the challenge request is detected... the session.get() will not solve anything as its a direct call to requests.... 

if you issue the cf.get() first, the session you passed is then populated with the cfuid cookie so you can use session.get() the problem with loading each thread with the existing cfscrape is your going to be importing js2py over and over again on each thread, which is extremely slow

**Comment by VeNoMouS** [comment]
@noembryo you will like the new project then no re-factoring imho ;P

**Comment by noembryo** [comment]
>  the problem with loading each thread cf.create_scraper() with the existing cfscrape is your going to be importing js2py over and over again on each thread, which is extremely slow

I don't do that.
I create one scraper in the search session and then pass this scraper to every thread.
Inside the every thread, there is a check to see if someone ask the site.
The first one does asks and blocks the other.
When the first one get its results, all the others start asking with the same object.

**Comment by noembryo** [comment]
> you will like the new project then no re-factoring imho ;P

Its a pity that until the time you'll have a working release, I'll have to have the refactory done :-p

**Comment by pro-src** [comment]
I think it's best to recommend that people continue to use `scraper.get` over `session.get` for the following reasons:

* scraper is a session already
* CF can send additional challenges after the cookie has been set before the cookie expires
* session isn't going to know whether the response contains a challenge
* I've noticed a pattern of getting an additional challenge after the first few requests with the cookie already set (probably due to CF's BIC) but seems random
* It over complicates things when the only real reason to use session is if you already have it or micro-optimization (negligible performance gains)

**Comment by VeNoMouS** [comment]
@noembryo it works atm... its up to you,.. we only have a dev branch atm..

**Comment by noembryo** [comment]
I don't mind to wait..
I'll refactor the code for the fun of it ;o)
I can always return to a previous state when you release it, no worries..
Thank you anyway :o)

**Comment by VeNoMouS** [comment]
All good mate :)

**Comment by pannal** [comment]
@VeNoMouS do you have a link to the new project? :)

**Comment by noembryo** [comment]
Xmm, I think there is a problem today..

Using `result = self.scraper.get(url).content` returns nothing.
Even for one file. Not even an error.

Is there something wrong with the CloudFlare or with my provider?

**Comment by pro-src** [comment]
Provider.

**Comment by noembryo** [comment]
After some testing, found out about the problem.
With *every* request it asks for Captcha.
Some with changed IP and Header..

No luck anymore.. :o(

**Comment by noembryo** [comment]
Just to check.. can you guys see if you can get anything from SubScene?
```python
import cfscrape
scraper = cfscrape.create_scraper()
print(scraper.get("https://subscene.com/").content)
```

**Comment by VeNoMouS** [comment]
Project has been superseded by my other project ["cloudscraper"](https://pypi.org/project/cloudscraper/) the new git repo can be found [here](https://github.com/VeNoMouS/cloudscraper).
